### PR TITLE
Rename GradMonitor -> OptimizerMonitor; add functionality to log optimizer-specific metrics to assist loss spike investigation

### PR DESCRIPTION
**Pull request by bmosaicml**
# What does this PR do?

1. Changes GradMonitor -> OptimizerMonitor
2. Add functionality to request optimizer-specific metrics via optimizers `report_per_parameter_metrics`
3. Add `report_per_parameter_metrics` to DecoupledAdamW that reports a bunch of cool metrics like per param grad norm, moment norms, angles b/w moments and gradients, etc.

# What issue(s) does this change relate to?

https://mosaicml.atlassian.net/browse/RESEARCH-314?atlOrigin=eyJpIjoiOTFmYWJkYTFlNWI4NDdkNTk1YWIxOTJjODVkYWFlYzMiLCJwIjoiaiJ9

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by bandish-shah — CHANGES_REQUESTED** [review]
Please fill out the PR details as requested by the template.  This should include a concise description of what this PR addresses and there needs to be a Jira issue that tracks it.  Additionally, please change the PR title to something descriptive, not the branch name.  This is important for tracking so we don't drop important details during release.

File: tests/callbacks/test_optimizer_monitor.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    in_memory_logger = InMemoryLogger()  # track the logged metrics in the in_memory_logger
+    model = SimpleModel()
+    # Construct the trainer and train
+    trainer = Trainer(
+        model=model,
+        callbacks=grad_monitor,
+        loggers=in_memory_logger,
+        train_dataloader=DataLoader(RandomClassificationDataset()),
+        optimizers=DecoupledAdamW(model.parameters()),
+        max_duration='2ep',
```
Comment (mvpatel2000) [inline tests/callbacks/test_optimizer_monitor.py:29]: do we need to run 2ep instead of 1ep like old test?

File: composer/callbacks/optimizer_monitor.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                    layer_norms[f'layer_grad_l2_norm/{name}'] = param_grad_norm
+
+                param_grad_norm = param_grad_norm**2
+                norm += param_grad_norm
+                metric_reporter = getattr(state.optimizers[0], 'report_per_parameter_metrics', None)
+                if callable(metric_reporter) and self.log_optimizer_metrics:
+                    optimizer_metrics = metric_reporter(p, name, optimizer_metrics)
+
+        norm = norm**0.5
+        logger.log_metrics({'grad_l2_norm/step': norm})
```
Comment (mvpatel2000) [inline composer/callbacks/optimizer_monitor.py:111]: nit
```suggestion
        logger.log_metrics({'grad_l2_norm/step': norm**0.5})
```

**Review by mvpatel2000 — COMMENTED** [review]
Mostly LGTM. Defer to Abhi for approval

File: composer/callbacks/optimizer_monitor.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            if p.grad is not None and p.requires_grad:
+                param_grad_norm = p.grad.detach().data.norm(2).item()  # type: ignore
+                if self.log_layer_grad_norms:
+                    layer_norms[f'layer_grad_l2_norm/{name}'] = param_grad_norm
+
+                param_grad_norm = param_grad_norm**2
+                norm += param_grad_norm
+                metric_reporter = getattr(state.optimizers[0], 'report_per_parameter_metrics', None)
+                if callable(metric_reporter) and self.log_optimizer_metrics:
+                    optimizer_metrics = metric_reporter(p, name, optimizer_metrics)
```
Comment (abhi-mosaic) [inline composer/callbacks/optimizer_monitor.py:108]: Right here is where we could add a prefix `optimizer_monitor/` to all metric keys

File: composer/callbacks/optimizer_monitor.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    |                                               | Layer-wise ratio between step size and parameter    |
+    | ``layer_step_param_norm_ratio/LAYER_NAME``    | norm                                                |
+    |                                               |                                                     |
+    +-----------------------------------------------+-----------------------------------------------------+
+
+    Args:
+        log_layer_grad_norms (bool, optional): Whether to log the L2 normalization of each layer.
+            Default: ``False``.
+        log_layer_grad_norms (bool, optional): Whether to log optimizer-specific metrics.
+            Default: ``False``.
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/callbacks/optimizer_monitor.py:86]: Can we refactor these "default metrics" in a cleaner way? Like, can we just put them into a `default_metric_reporter` that gets merged with any additional metrics we find on the optimizer?

**Review by abhi-mosaic — COMMENTED** [review]
Looking pretty good! I made some comments, mostly about logging names and cleaning up the default metrics (grad_norm, layer_grad_norm) a bit. Thank you for this amazing work @bmosaicml !

Only potential bugs are the `sqrt()` naming, and the lack of weight decay in the manual `step_tensor` that you are creating

File: composer/callbacks/optimizer_monitor.py (hunk 3 above)
Comment (bmosaicml) [reply composer/callbacks/optimizer_monitor.py:101]: I tried this out but the UX ended up feeling a little off having the default metrics (grad norms) separated from the optimizer metrics (moment norms etc.) since they are all conceptually related. I did end up following your suggesting of grouping the metrics hierarchically by type!

File: tests/callbacks/test_optimizer_monitor.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    in_memory_logger = InMemoryLogger()  # track the logged metrics in the in_memory_logger
+    model = SimpleModel()
+    # Construct the trainer and train
+    trainer = Trainer(
+        model=model,
+        callbacks=grad_monitor,
+        loggers=in_memory_logger,
+        train_dataloader=DataLoader(RandomClassificationDataset()),
+        optimizers=DecoupledAdamW(model.parameters()),
+        max_duration='1ep',
```
Comment (mvpatel2000) [inline tests/callbacks/test_optimizer_monitor.py:28]: nit: since this logs per batch, can we instead set max duration to `3ba` or something? 

We're cleaning up our slow tests to try to speedup CI/CD workflow, and smaller fit runs are always helpful

**Review by mvpatel2000 — APPROVED** [review]
LGTM mod minor nits. I'm approving since I'll be OOO and this needs an eng stamp, but it should not be merged until someone from research / @abhi-mosaic gives sign off as well

File: tests/callbacks/test_optimizer_monitor.py (hunk 5 above)
Comment (bmosaicml) [reply tests/callbacks/test_optimizer_monitor.py]: good suggestions!

**Review by abhi-mosaic — APPROVED** [review]
LGTM! Thank you @bmosaicml for addressing all the prev comments. 

I think we should log a low priority JIRA ticket for the weight decay issue: currently if `weight_decay` is on, the `step_tensor` in is not calculated exactly right because we don't have access to the weights before the update, on line 369 in `decoupled_weight_decay.py`. But it looks good to merge and already seems to be helping out with NLP loss spike research!
